### PR TITLE
Fixed #25388 -- Added an option to disable migrations when running tests.

### PR DESCRIPTION
**Pull request by dadoeyad**
https://code.djangoproject.com/ticket/25388

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, add to whitelist.

**Comment by tobiasmcnulty** [comment]
Please squash into a single commit with a message per the guidelines here: https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines

Since you are working on top of someone else's patch, it would also be nice to keep the original Git author or at least include a "Thanks to ... for the patch." in the commit message.

**Comment by tobiasmcnulty** [comment]
Some proposed edits to the settings docs:

	diff --git a/docs/ref/settings.txt b/docs/ref/settings.txt
	index 8d5f368..6f19cc8 100644
	--- a/docs/ref/settings.txt
	+++ b/docs/ref/settings.txt
	@@ -707,8 +707,13 @@ on :ref:`controlling the creation order of test databases
	 
	 Default: ``True``
	 
	-If it is set to ``False``, Django won't use existing migrations to create a
	-test database.
	+Boolean value to control whether or not migrations are used to create the test
	+database.
	+
	+If it is set to ``False``, Django will create the test database using app
	+models directly rather than by applying migrations. Note that disabling this
	+option will skip *all* migrations when running tests, including any data
	+migrations.
	 
	 .. setting:: TEST_MIRROR
	 
	@@ -1923,10 +1928,6 @@ automatically create the package if it doesn't already exist.
	 
	 When you supply ``None`` as a value for an app, Django will consider the app as
	 an app without migrations regardless of an existing ``migrations`` submodule.
	-This can be used, for example, in a test settings file to skip migrations while
	-testing (tables will still be created for the apps' models). If this is used in
	-your general project settings, remember to use the :option:`migrate
	---run-syncdb` option if you want to create tables for the app.
	 
	 .. setting:: MONTH_DAY_FORMAT

**Comment by tobiasmcnulty** [comment]
More importantly, I noticed that this option and `--keepdb` are incompatible:

    ProgrammingError: column "test2" of relation "bar_foo" does not exist
    LINE 1: INSERT INTO "bar_foo" ("bar", "test", "test2") VALUES ('test...

I.e., if you have `'MIGRATE': False` and pass `--keepdb`, Django can't update your database and it won't be recreated. This isn't necessarily a problem, but it should be documented (e.g., in `topics/testing/overview.txt` and `ref/django-admin.txt`, along with a link to the `TEST_MIGRATE` setting reference for details).

**Comment by tobiasmcnulty** [comment]
Aside from the above I did test this and confirmed that it successfully disables migrations when running tests and does NOT disable migrations when simply calling `manage.py migrate` (the reason the previous patch was reverted).

IMHO it would be nice to find a way to test this (and that normal migrations are unaffected) automatically, even if it requires some mocking.

File: django/db/backends/base/creation.py [hunk 1]
```diff
@@ -59,6 +59,8 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
         settings.DATABASES[self.connection.alias]["NAME"] = test_database_name
         self.connection.settings_dict["NAME"] = test_database_name
 
+        self.connection._run_in_test_case = True
```
Comment (tobiasmcnulty) [inline django/db/backends/base/creation.py]: If we keep this approach, it's likely to be adopted elsewhere so we should choose the name carefully. That said, I'm relatively ambivalent about what it should be -- any opinions?

**Comment by LuRsT** [comment]
Hey @dadoeyad, need help with this PR?

**Comment by dadoeyad** [comment]
> Hey @dadoeyad, need help with this PR?

Sure. You can take charge of it if you'd like :) I haven't worked on it in years.

**Comment by felixxm** [comment]
Superseded by #12062

**Pull request closed without merging**